In [0]:
# ========================================
# EXPLORATORY NOTEBOOK
# DATASET: gold_mart_sales_performance
# ========================================

# ========================================
# 0. CONFIGURATION
# ========================================

CATALOG = "ptfrozenfoods_dev"
SOURCE_SCHEMA = "gold"
TARGET_SCHEMA = "gold"

DOMAIN = "marts"
DATASET = "gold_mart_sales_performance"

STORAGE_ACCOUNT = "stptfrozenfoodsdevwe01"
GOLD_CONTAINER = "gold"

FACT_TABLE_NAME = "fact_sales"
DIM_CALENDAR_TABLE_NAME = "dim_calendar"
DIM_PRODUCT_TABLE_NAME = "dim_product"
DIM_CUSTOMER_TABLE_NAME = "dim_customer"
DIM_SUPPLIER_TABLE_NAME = "dim_supplier"
DIM_SALES_CHANNEL_TABLE_NAME = "dim_sales_channel"

FACT_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{FACT_TABLE_NAME}"
DIM_CALENDAR_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{DIM_CALENDAR_TABLE_NAME}"
DIM_PRODUCT_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{DIM_PRODUCT_TABLE_NAME}"
DIM_CUSTOMER_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{DIM_CUSTOMER_TABLE_NAME}"
DIM_SUPPLIER_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{DIM_SUPPLIER_TABLE_NAME}"
DIM_SALES_CHANNEL_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{DIM_SALES_CHANNEL_TABLE_NAME}"

CANDIDATE_TARGET_TABLE = f"{CATALOG}.{TARGET_SCHEMA}.mart_sales_performance"

GOLD_BASE_PATH = f"abfss://{GOLD_CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/"

FACT_SOURCE_PATH = f"{GOLD_BASE_PATH}facts/fact_sales/"
DIM_CALENDAR_SOURCE_PATH = f"{GOLD_BASE_PATH}dimensions/dim_calendar/"
DIM_PRODUCT_SOURCE_PATH = f"{GOLD_BASE_PATH}dimensions/dim_product/"
DIM_CUSTOMER_SOURCE_PATH = f"{GOLD_BASE_PATH}dimensions/dim_customer/"
DIM_SUPPLIER_SOURCE_PATH = f"{GOLD_BASE_PATH}dimensions/dim_supplier/"
DIM_SALES_CHANNEL_SOURCE_PATH = f"{GOLD_BASE_PATH}dimensions/dim_sales_channel/"

CANDIDATE_TARGET_PATH = f"{GOLD_BASE_PATH}marts/mart_sales_performance/"

In [0]:
# ========================================
# 1. CONTEXT SETUP
# ========================================

print("=" * 80)
print("STARTING GOLD EXPLORATORY: gold_mart_sales_performance")
print("=" * 80)

spark.sql(f"USE CATALOG {CATALOG}")
spark.sql(f"USE SCHEMA {SOURCE_SCHEMA}")

print("[INFO] Context setup completed successfully.")

STARTING GOLD EXPLORATORY: gold_mart_sales_performance
[INFO] Context setup completed successfully.


In [0]:
# ========================================
# 2. CONFIGURATION SUMMARY
# ========================================

print("=" * 80)
print("GOLD EXPLORATORY NOTEBOOK CONFIGURATION")
print("=" * 80)
print(f"Catalog:                         {CATALOG}")
print(f"Source schema:                   {SOURCE_SCHEMA}")
print(f"Target schema:                   {TARGET_SCHEMA}")
print(f"Domain:                          {DOMAIN}")
print(f"Dataset:                         {DATASET}")
print(f"Fact table:                      {FACT_TABLE}")
print(f"Dim calendar table:              {DIM_CALENDAR_TABLE}")
print(f"Dim product table:               {DIM_PRODUCT_TABLE}")
print(f"Dim customer table:              {DIM_CUSTOMER_TABLE}")
print(f"Dim supplier table:              {DIM_SUPPLIER_TABLE}")
print(f"Dim sales channel table:         {DIM_SALES_CHANNEL_TABLE}")
print(f"Candidate target table:          {CANDIDATE_TARGET_TABLE}")
print(f"Fact source path:                {FACT_SOURCE_PATH}")
print(f"Candidate target path:           {CANDIDATE_TARGET_PATH}")
print("=" * 80)

GOLD EXPLORATORY NOTEBOOK CONFIGURATION
Catalog:                         ptfrozenfoods_dev
Source schema:                   gold
Target schema:                   gold
Domain:                          marts
Dataset:                         gold_mart_sales_performance
Fact table:                      ptfrozenfoods_dev.gold.fact_sales
Dim calendar table:              ptfrozenfoods_dev.gold.dim_calendar
Dim product table:               ptfrozenfoods_dev.gold.dim_product
Dim customer table:              ptfrozenfoods_dev.gold.dim_customer
Dim supplier table:              ptfrozenfoods_dev.gold.dim_supplier
Dim sales channel table:         ptfrozenfoods_dev.gold.dim_sales_channel
Candidate target table:          ptfrozenfoods_dev.gold.mart_sales_performance
Fact source path:                abfss://gold@stptfrozenfoodsdevwe01.dfs.core.windows.net/facts/fact_sales/
Candidate target path:           abfss://gold@stptfrozenfoodsdevwe01.dfs.core.windows.net/marts/mart_sales_performance/


In [0]:
# ========================================
# 3. PRE-CHECKS
# ========================================

print("[INFO] Checking source table availability...")
spark.sql(f"DESCRIBE TABLE {FACT_TABLE}")
spark.sql(f"DESCRIBE TABLE {DIM_CALENDAR_TABLE}")
spark.sql(f"DESCRIBE TABLE {DIM_PRODUCT_TABLE}")
spark.sql(f"DESCRIBE TABLE {DIM_CUSTOMER_TABLE}")
spark.sql(f"DESCRIBE TABLE {DIM_SUPPLIER_TABLE}")
spark.sql(f"DESCRIBE TABLE {DIM_SALES_CHANNEL_TABLE}")

print("[INFO] Checking source path access...")
dbutils.fs.ls(FACT_SOURCE_PATH)
dbutils.fs.ls(DIM_CALENDAR_SOURCE_PATH)
dbutils.fs.ls(DIM_PRODUCT_SOURCE_PATH)
dbutils.fs.ls(DIM_CUSTOMER_SOURCE_PATH)
dbutils.fs.ls(DIM_SUPPLIER_SOURCE_PATH)
dbutils.fs.ls(DIM_SALES_CHANNEL_SOURCE_PATH)

print("[INFO] Checking target container access...")
dbutils.fs.ls(f"abfss://{GOLD_CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/")

print("[INFO] Pre-checks completed successfully.")

[INFO] Checking source table availability...
[INFO] Checking source path access...
[INFO] Checking target container access...
[INFO] Pre-checks completed successfully.


In [0]:
# ========================================
# 4. READ SOURCE DATA
# ========================================

print("[INFO] Loading Gold source tables...")

df_fact_sales = spark.table(FACT_TABLE)
df_dim_calendar = spark.table(DIM_CALENDAR_TABLE)
df_dim_product = spark.table(DIM_PRODUCT_TABLE)
df_dim_customer = spark.table(DIM_CUSTOMER_TABLE)
df_dim_supplier = spark.table(DIM_SUPPLIER_TABLE)
df_dim_sales_channel = spark.table(DIM_SALES_CHANNEL_TABLE)

print("[INFO] Source tables loaded successfully.")

print("=" * 80)
print("SOURCE DATA ROW COUNTS")
print("=" * 80)
print(f"Fact sales:          {df_fact_sales.count():,}")
print(f"Dim calendar:        {df_dim_calendar.count():,}")
print(f"Dim product:         {df_dim_product.count():,}")
print(f"Dim customer:        {df_dim_customer.count():,}")
print(f"Dim supplier:        {df_dim_supplier.count():,}")
print(f"Dim sales channel:   {df_dim_sales_channel.count():,}")
print("=" * 80)

[INFO] Loading Gold source tables...
[INFO] Source tables loaded successfully.
SOURCE DATA ROW COUNTS
Fact sales:          244,357
Dim calendar:        1,885
Dim product:         36
Dim customer:        145
Dim supplier:        6
Dim sales channel:   4


In [0]:
# ========================================
# DATASET PREVIEW — INITIAL EXPLORATION
# ========================================

print("=" * 80)
print("DATASET PREVIEW — INITIAL EXPLORATION")
print("=" * 80)

print("\n[INFO] Preview: FACT SALES (df_fact_sales)")
print("-" * 80)
display(df_fact_sales.limit(5))

print("\n[INFO] Preview: DIM CALENDAR (df_dim_calendar)")
print("-" * 80)
display(df_dim_calendar.limit(5))

print("\n[INFO] Preview: DIM PRODUCT (df_dim_product)")
print("-" * 80)
display(df_dim_product.limit(5))

print("\n[INFO] Preview: DIM CUSTOMER (df_dim_customer)")
print("-" * 80)
display(df_dim_customer.limit(5))

print("\n[INFO] Preview: DIM SUPPLIER (df_dim_supplier)")
print("-" * 80)
display(df_dim_supplier.limit(5))

print("\n[INFO] Preview: DIM SALES CHANNEL (df_dim_sales_channel)")
print("-" * 80)
display(df_dim_sales_channel.limit(5))

print("\n[INFO] Printing schemas...")
df_fact_sales.printSchema()
df_dim_calendar.printSchema()
df_dim_product.printSchema()
df_dim_customer.printSchema()
df_dim_supplier.printSchema()
df_dim_sales_channel.printSchema()

print("\n" + "=" * 80)
print("[INFO] Dataset preview completed successfully.")
print("=" * 80)

DATASET PREVIEW — INITIAL EXPLORATION

[INFO] Preview: FACT SALES (df_fact_sales)
--------------------------------------------------------------------------------


item_pedido_id,pedido_id,produto_id,cliente_id,canal_id,fornecedor_id,data_pedido,calendar_year,calendar_quarter,calendar_month,calendar_day,calendar_month_name,calendar_day_of_week,calendar_day_of_week_name,calendar_is_weekend,calendar_is_month_start,calendar_is_month_end,vendedor_id,cidade_entrega,estado_pedido,prazo_entrega_dias,tipo_cliente,cliente_cidade,distrito,segmento,potencial_valor,cluster_comercial,status_cliente,produto_nome,categoria,marca,status_produto,nome_canal,descricao_canal,quantity_sold,preco_lista_unitario,desconto_unitario,preco_venda_unitario,custo_unitario,gross_sales_amount,net_sales_amount,total_cost_amount,gross_margin_amount,line_count,flag_promocao,lote_fornecedor,item_load_date,item_ingestion_timestamp,item_source_file,order_load_date,order_ingestion_timestamp,order_source_file,average_order_value
IT000000027,PED2021011100012,P012,C0136,CH04,F006,2021-01-11,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,UNKNOWN,Espinho,Entregue,1,Restaurante,Espinho,Aveiro,Pequeno,Alto,Cluster C,Ativo,Hambúrguer bovino 800g,Carne,Chef Express,Ativo,Marketplace,Plataformas externas de venda,2,5.49,0.0,5.49,3.14,10.98,10.98,6.28,4.7,1,0,L66962,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,2026-03-17,2026-04-03T22:23:23.76Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,29.42
IT000000026,PED2021011100012,P020,C0136,CH04,F002,2021-01-11,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,UNKNOWN,Espinho,Entregue,1,Restaurante,Espinho,Aveiro,Pequeno,Alto,Cluster C,Ativo,Batata smile 1500g,Batatas,FrioMax,Ativo,Marketplace,Plataformas externas de venda,4,4.61,0.0,4.61,2.34,18.44,18.44,9.36,9.080000000000002,1,0,L39662,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,2026-03-17,2026-04-03T22:23:23.76Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,29.42
IT000000028,PED2021011100013,P004,C0136,CH02,F001,2021-01-11,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,VEND001,Espinho,Entregue,1,Restaurante,Espinho,Aveiro,Pequeno,Alto,Cluster C,Ativo,Bacalhau desfiado 800g,Peixe,PT Frozen Foods,Ativo,Vendas Externas,Equipa comercial visitando clientes,2,8.72,0.0,8.72,4.9,17.44,17.44,9.8,7.640000000000001,1,0,L28888,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,2026-03-17,2026-04-03T22:23:23.76Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,53.79
IT000000030,PED2021011100013,P023,C0136,CH02,F005,2021-01-11,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,VEND001,Espinho,Entregue,1,Restaurante,Espinho,Aveiro,Pequeno,Alto,Cluster C,Ativo,Pão de forma 800g,Padaria,Doce Norte,Ativo,Vendas Externas,Equipa comercial visitando clientes,4,2.97,0.15,2.82,1.36,11.88,11.28,5.44,5.839999999999999,1,1,L45385,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,2026-03-17,2026-04-03T22:23:23.76Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,53.79
IT000000029,PED2021011100013,P020,C0136,CH02,F002,2021-01-11,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,VEND001,Espinho,Entregue,1,Restaurante,Espinho,Aveiro,Pequeno,Alto,Cluster C,Ativo,Batata smile 1500g,Batatas,FrioMax,Ativo,Vendas Externas,Equipa comercial visitando clientes,3,4.93,0.0,4.93,2.24,14.79,14.79,6.720000000000001,8.069999999999999,1,0,L68711,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,2026-03-17,2026-04-03T22:23:23.76Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.c


[INFO] Preview: DIM CALENDAR (df_dim_calendar)
--------------------------------------------------------------------------------


calendar_date,calendar_year,calendar_month,calendar_day,calendar_quarter,calendar_month_name,calendar_day_of_week,calendar_day_name,calendar_is_weekend,calendar_is_month_start,calendar_is_month_end
2021-01-13,2021,1,13,1,Janeiro,3,Quarta-feira,0,0,0
2021-01-16,2021,1,16,1,Janeiro,6,Sábado,1,0,0
2021-01-27,2021,1,27,1,Janeiro,3,Quarta-feira,0,0,1
2021-02-04,2021,2,4,1,Fevereiro,4,Quinta-feira,0,1,0
2021-04-05,2021,4,5,2,Abril,1,Segunda-feira,0,1,0



[INFO] Preview: DIM PRODUCT (df_dim_product)
--------------------------------------------------------------------------------


produto_id,produto_nome,categoria,marca,peso_gramas,preco_lista_base,custo_base_unitario,fornecedor_id,data_lancamento,data_fim,status_produto,popularidade_base,sensibilidade_promocao,fator_sazonal_proprio
P032,"Ice tea limão 1,5L 1500g",Bebidas,Chef Express,1500,1.7,0.86,F006,2021-11-13,2024-03-20,Inativo,4.243,0.929,1.064
P035,Ervilhas em lata 400g,Conservas & Secos,Campo Verde,400,0.9,0.45,F004,2022-09-12,null,Ativo,1.511,0.676,0.782
P012,Hambúrguer bovino 800g,Carne,Chef Express,800,5.9,3.32,F006,2022-04-12,null,Ativo,2.695,0.557,0.922
P016,Empadão de carne 1200g,Pré-cozinhados,Atlântico Select,1200,7.1,4.3,F001,2023-12-06,null,Ativo,2.871,0.982,0.975
P007,Cocktail de marisco 1kg,Marisco,FrioMax,1000,9.8,6.35,F002,2023-09-03,null,Ativo,1.196,1.259,1.155



[INFO] Preview: DIM CUSTOMER (df_dim_customer)
--------------------------------------------------------------------------------


cliente_id,tipo_cliente,segmento,cliente_cidade,distrito,potencial_valor,cluster_comercial,status_cliente
C0059,Restaurante,Pequeno,Valongo,Porto,Baixo,Cluster C,Ativo
C0195,Hotel,Grande,Porto,Porto,Médio,Cluster B,Ativo
C0167,Particular,Pequeno,Vila Nova de Gaia,Porto,Baixo,Cluster B,Ativo
C0150,Take-away,Médio,Barcelos,Braga,Médio,Cluster B,Ativo
C0016,Supermercado,Pequeno,Valongo,Porto,Médio,Cluster C,Ativo



[INFO] Preview: DIM SUPPLIER (df_dim_supplier)
--------------------------------------------------------------------------------


fornecedor_id,nome_fornecedor,pais,status_fornecedor,codigo_legacy,ultima_sincronizacao,load_date,ingestion_timestamp,source_file
F003,PortoMar Congelados,Portugal,Ativo,LEG-F102,2026-03-15T22:16:51Z,2026-03-17,2026-04-03T22:51:50.272Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_suppliers/load_date=2026-03-17/erp_suppliers.csv
F005,Doces do Norte,Portugal,Ativo,LEG-F104,2026-03-15T22:16:51Z,2026-03-17,2026-04-03T22:51:50.272Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_suppliers/load_date=2026-03-17/erp_suppliers.csv
F006,Mar & Serra Foods,Portugal,Ativo,LEG-F105,2026-03-15T22:16:51Z,2026-03-17,2026-04-03T22:51:50.272Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_suppliers/load_date=2026-03-17/erp_suppliers.csv
F004,DouroHortícola,Portugal,Ativo,LEG-F103,2026-03-15T22:16:51Z,2026-03-17,2026-04-03T22:51:50.272Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_suppliers/load_date=2026-03-17/erp_suppliers.csv
F002,Atlântico Frio Distribuição,Portugal,Ativo,LEG-F101,2026-03-15T22:16:51Z,2026-03-17,2026-04-03T22:51:50.272Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_suppliers/load_date=2026-03-17/erp_suppliers.csv



[INFO] Preview: DIM SALES CHANNEL (df_dim_sales_channel)
--------------------------------------------------------------------------------


canal_id,nome_canal,descricao
CH01,E-commerce,Pedidos realizados pelo site B2B/B2C
CH04,Marketplace,Plataformas externas de venda
CH02,Vendas Externas,Equipa comercial visitando clientes
CH03,Telefone,Pedidos feitos por telefone



[INFO] Printing schemas...
root
 |-- item_pedido_id: string (nullable = true)
 |-- pedido_id: string (nullable = true)
 |-- produto_id: string (nullable = true)
 |-- cliente_id: string (nullable = true)
 |-- canal_id: string (nullable = true)
 |-- fornecedor_id: string (nullable = true)
 |-- data_pedido: date (nullable = true)
 |-- calendar_year: integer (nullable = true)
 |-- calendar_quarter: integer (nullable = true)
 |-- calendar_month: integer (nullable = true)
 |-- calendar_day: integer (nullable = true)
 |-- calendar_month_name: string (nullable = true)
 |-- calendar_day_of_week: integer (nullable = true)
 |-- calendar_day_of_week_name: string (nullable = true)
 |-- calendar_is_weekend: integer (nullable = true)
 |-- calendar_is_month_start: integer (nullable = true)
 |-- calendar_is_month_end: integer (nullable = true)
 |-- vendedor_id: string (nullable = true)
 |-- cidade_entrega: string (nullable = true)
 |-- estado_pedido: string (nullable = true)
 |-- prazo_entrega_dias: i

In [0]:
# ========================================
# 5. SOURCE VALIDATION
# ========================================

from pyspark.sql import functions as F

required_fact_columns = [
    "data_pedido",
    "produto_id",
    "cliente_id",
    "fornecedor_id",
    "canal_id",
    "quantity_sold",
    "gross_sales_amount",
    "net_sales_amount",
    "total_cost_amount",
    "gross_margin_amount",
    "line_count",
    "average_order_value"
]

required_calendar_columns = [
    "calendar_date",
    "calendar_year",
    "calendar_quarter",
    "calendar_month",
    "calendar_day"
]

required_product_columns = [
    "produto_id",
    "produto_nome",
    "categoria",
    "marca",
    "fornecedor_id",
    "status_produto"
]

required_customer_columns = [
    "cliente_id",
    "tipo_cliente",
    "segmento",
    "cliente_cidade",
    "distrito",
    "status_cliente"
]

required_supplier_columns = [
    "fornecedor_id",
    "nome_fornecedor",
    "pais",
    "status_fornecedor"
]

required_sales_channel_columns = [
    "canal_id",
    "nome_canal",
    "descricao"
]

missing_fact_columns = [c for c in required_fact_columns if c not in df_fact_sales.columns]
missing_calendar_columns = [c for c in required_calendar_columns if c not in df_dim_calendar.columns]
missing_product_columns = [c for c in required_product_columns if c not in df_dim_product.columns]
missing_customer_columns = [c for c in required_customer_columns if c not in df_dim_customer.columns]
missing_supplier_columns = [c for c in required_supplier_columns if c not in df_dim_supplier.columns]
missing_sales_channel_columns = [c for c in required_sales_channel_columns if c not in df_dim_sales_channel.columns]

print(f"[INFO] Missing fact columns: {missing_fact_columns}")
print(f"[INFO] Missing calendar columns: {missing_calendar_columns}")
print(f"[INFO] Missing product columns: {missing_product_columns}")
print(f"[INFO] Missing customer columns: {missing_customer_columns}")
print(f"[INFO] Missing supplier columns: {missing_supplier_columns}")
print(f"[INFO] Missing sales channel columns: {missing_sales_channel_columns}")

if (
    missing_fact_columns
    or missing_calendar_columns
    or missing_product_columns
    or missing_customer_columns
    or missing_supplier_columns
    or missing_sales_channel_columns
):
    raise Exception("Missing required columns in one or more source tables.")

print("[INFO] Source validation completed successfully.")

[INFO] Missing fact columns: []
[INFO] Missing calendar columns: []
[INFO] Missing product columns: []
[INFO] Missing customer columns: []
[INFO] Missing supplier columns: []
[INFO] Missing sales channel columns: []
[INFO] Source validation completed successfully.


In [0]:
# ========================================
# 6. INITIAL DATA QUALITY CHECKS
# ========================================

display(
    df_fact_sales.select(
        F.count("*").alias("total_rows"),
        F.countDistinct("data_pedido").alias("distinct_data_pedido"),
        F.countDistinct("produto_id").alias("distinct_produto_id"),
        F.countDistinct("cliente_id").alias("distinct_cliente_id"),
        F.countDistinct("fornecedor_id").alias("distinct_fornecedor_id"),
        F.countDistinct("canal_id").alias("distinct_canal_id"),
        F.sum(F.when(F.col("data_pedido").isNull(), 1).otherwise(0)).alias("null_data_pedido"),
        F.sum(F.when(F.col("produto_id").isNull(), 1).otherwise(0)).alias("null_produto_id"),
        F.sum(F.when(F.col("cliente_id").isNull(), 1).otherwise(0)).alias("null_cliente_id"),
        F.sum(F.when(F.col("fornecedor_id").isNull(), 1).otherwise(0)).alias("null_fornecedor_id"),
        F.sum(F.when(F.col("canal_id").isNull(), 1).otherwise(0)).alias("null_canal_id"),
        F.sum(F.when(F.col("net_sales_amount").isNull(), 1).otherwise(0)).alias("null_net_sales_amount"),
        F.sum(F.when(F.col("gross_margin_amount").isNull(), 1).otherwise(0)).alias("null_gross_margin_amount")
    )
)

total_rows,distinct_data_pedido,distinct_produto_id,distinct_cliente_id,distinct_fornecedor_id,distinct_canal_id,null_data_pedido,null_produto_id,null_cliente_id,null_fornecedor_id,null_canal_id,null_net_sales_amount,null_gross_margin_amount
244357,1875,36,145,6,4,0,0,0,0,0,0,0


In [0]:
# ========================================
# 7. REFERENTIAL INTEGRITY CHECKS
# ========================================

print("[INFO] Checking referential integrity between fact_sales and dimensions...")

calendar_missing = (
    df_fact_sales.select("data_pedido").distinct()
    .join(
        df_dim_calendar.select(F.col("calendar_date").alias("data_pedido")).distinct(),
        on="data_pedido",
        how="left_anti"
    )
)

product_missing = (
    df_fact_sales.select("produto_id").distinct()
    .join(df_dim_product.select("produto_id").distinct(), on="produto_id", how="left_anti")
)

customer_missing = (
    df_fact_sales.select("cliente_id").distinct()
    .join(df_dim_customer.select("cliente_id").distinct(), on="cliente_id", how="left_anti")
)

supplier_missing = (
    df_fact_sales.select("fornecedor_id").distinct()
    .join(df_dim_supplier.select("fornecedor_id").distinct(), on="fornecedor_id", how="left_anti")
)

sales_channel_missing = (
    df_fact_sales.select("canal_id").distinct()
    .join(df_dim_sales_channel.select("canal_id").distinct(), on="canal_id", how="left_anti")
)

print(f"[INFO] Missing calendar keys:       {calendar_missing.count():,}")
print(f"[INFO] Missing product keys:        {product_missing.count():,}")
print(f"[INFO] Missing customer keys:       {customer_missing.count():,}")
print(f"[INFO] Missing supplier keys:       {supplier_missing.count():,}")
print(f"[INFO] Missing sales channel keys:  {sales_channel_missing.count():,}")

[INFO] Checking referential integrity between fact_sales and dimensions...
[INFO] Missing calendar keys:       0
[INFO] Missing product keys:        0
[INFO] Missing customer keys:       0
[INFO] Missing supplier keys:       0
[INFO] Missing sales channel keys:  0


In [0]:
# ========================================
# 8. BUSINESS PROFILE CHECKS
# ========================================

print("[INFO] Net sales by sales channel")
display(
    df_fact_sales.groupBy("canal_id")
    .agg(
        F.round(F.sum("net_sales_amount"), 2).alias("total_net_sales"),
        F.round(F.sum("gross_margin_amount"), 2).alias("total_gross_margin"),
        F.sum("quantity_sold").alias("total_quantity_sold"),
        F.countDistinct("pedido_id").alias("total_orders")
    )
    .orderBy(F.desc("total_net_sales"))
)

print("[INFO] Net sales by customer")
display(
    df_fact_sales.groupBy("cliente_id")
    .agg(
        F.round(F.sum("net_sales_amount"), 2).alias("total_net_sales"),
        F.countDistinct("pedido_id").alias("total_orders"),
        F.sum("quantity_sold").alias("total_quantity_sold")
    )
    .orderBy(F.desc("total_net_sales"))
)

print("[INFO] Net sales by product")
display(
    df_fact_sales.groupBy("produto_id")
    .agg(
        F.round(F.sum("net_sales_amount"), 2).alias("total_net_sales"),
        F.sum("quantity_sold").alias("total_quantity_sold"),
        F.countDistinct("pedido_id").alias("total_orders")
    )
    .orderBy(F.desc("total_net_sales"))
)

print("[INFO] Net sales by date")
display(
    df_fact_sales.groupBy("data_pedido")
    .agg(
        F.round(F.sum("net_sales_amount"), 2).alias("total_net_sales"),
        F.round(F.sum("gross_margin_amount"), 2).alias("total_gross_margin"),
        F.countDistinct("pedido_id").alias("total_orders"),
        F.sum("quantity_sold").alias("total_quantity_sold")
    )
    .orderBy("data_pedido")
)

[INFO] Net sales by sales channel


canal_id,total_net_sales,total_gross_margin,total_quantity_sold,total_orders
CH02,2308048.52,905638.1,388761,36931
CH03,1340274.2,525699.98,221946,22249
CH01,1250732.56,494887.74,207268,20614
CH04,588283.98,233016.95,96818,9732


[INFO] Net sales by customer


cliente_id,total_net_sales,total_orders,total_quantity_sold
C0164,379267.11,4640,58090
C0072,268120.63,3264,50331
C0260,245366.82,2971,40651
C0314,237682.05,3076,37234
C0103,224874.99,2981,43001
C0031,219373.69,2639,35372
C0043,213785.07,2139,38796
C0227,177809.19,2815,28636
C0139,177150.45,2241,34414
C0057,173566.39,3876,26536


[INFO] Net sales by product


produto_id,total_net_sales,total_quantity_sold,total_orders
P004,1038249.71,121966,25720
P007,541910.89,55859,13691
P015,500059.9,73042,13487
P006,401121.48,53694,12048
P012,365556.18,61862,16958
P020,301340.79,63635,16016
P014,210256.47,34664,8243
P005,205950.54,18773,5477
P017,201954.37,22567,5801
P003,173512.32,24988,6566


[INFO] Net sales by date


data_pedido,total_net_sales,total_gross_margin,total_orders,total_quantity_sold
2021-01-11,1256.6,486.31,35,192
2021-01-12,1103.31,421.07,39,167
2021-01-13,1786.13,715.15,43,274
2021-01-14,882.65,340.37,29,128
2021-01-15,1863.14,738.14,44,275
2021-01-16,1083.24,419.56,25,164
2021-01-17,992.56,374.99,24,140
2021-01-18,1096.47,429.55,34,163
2021-01-19,1961.54,767.97,46,296
2021-01-20,1515.41,590.22,42,234


In [0]:
# ========================================
# 9. BUILD CANDIDATE MART
# ========================================

df_mart_sales_performance_candidate = (
    df_fact_sales.alias("f")
    .join(
        df_dim_calendar.alias("c"),
        F.col("f.data_pedido") == F.col("c.calendar_date"),
        how="left"
    )
    .join(
        df_dim_product.alias("p"),
        F.col("f.produto_id") == F.col("p.produto_id"),
        how="left"
    )
    .join(
        df_dim_customer.alias("cu"),
        F.col("f.cliente_id") == F.col("cu.cliente_id"),
        how="left"
    )
    .join(
        df_dim_supplier.alias("s"),
        F.col("f.fornecedor_id") == F.col("s.fornecedor_id"),
        how="left"
    )
    .join(
        df_dim_sales_channel.alias("sc"),
        F.col("f.canal_id") == F.col("sc.canal_id"),
        how="left"
    )
    .select(
        F.col("f.item_pedido_id"),
        F.col("f.pedido_id"),
        F.col("f.data_pedido"),
        F.col("f.produto_id"),
        F.col("f.cliente_id"),
        F.col("f.fornecedor_id"),
        F.col("f.canal_id"),
        F.col("f.quantity_sold"),
        F.col("f.gross_sales_amount"),
        F.col("f.net_sales_amount"),
        F.col("f.total_cost_amount"),
        F.col("f.gross_margin_amount"),
        F.col("f.line_count"),
        F.col("f.average_order_value"),
        F.col("c.calendar_year"),
        F.col("c.calendar_quarter"),
        F.col("c.calendar_month"),
        F.col("c.calendar_day"),
        F.col("c.calendar_month_name"),
        F.col("c.calendar_day_of_week"),
        F.col("c.calendar_day_name"),
        F.col("c.calendar_is_weekend"),
        F.col("c.calendar_is_month_start"),
        F.col("c.calendar_is_month_end"),
        F.col("p.produto_nome"),
        F.col("p.categoria"),
        F.col("p.marca"),
        F.col("p.status_produto"),
        F.col("cu.tipo_cliente"),
        F.col("cu.segmento"),
        F.col("cu.cliente_cidade"),
        F.col("cu.distrito"),
        F.col("cu.potencial_valor"),
        F.col("cu.cluster_comercial"),
        F.col("cu.status_cliente"),
        F.col("s.nome_fornecedor"),
        F.col("s.pais").alias("fornecedor_pais"),
        F.col("s.status_fornecedor"),
        F.col("sc.nome_canal"),
        F.col("sc.descricao").alias("descricao_canal")
    )
)

print(f"[INFO] Candidate mart row count before aggregation: {df_mart_sales_performance_candidate.count():,}")

display(df_mart_sales_performance_candidate.limit(10))

[INFO] Candidate mart row count before aggregation: 244,357


item_pedido_id,pedido_id,data_pedido,produto_id,cliente_id,fornecedor_id,canal_id,quantity_sold,gross_sales_amount,net_sales_amount,total_cost_amount,gross_margin_amount,line_count,average_order_value,calendar_year,calendar_quarter,calendar_month,calendar_day,calendar_month_name,calendar_day_of_week,calendar_day_name,calendar_is_weekend,calendar_is_month_start,calendar_is_month_end,produto_nome,categoria,marca,status_produto,tipo_cliente,segmento,cliente_cidade,distrito,potencial_valor,cluster_comercial,status_cliente,nome_fornecedor,fornecedor_pais,status_fornecedor,nome_canal,descricao_canal
IT000000027,PED2021011100012,2021-01-11,P012,C0136,F006,CH04,2,10.98,10.98,6.28,4.7,1,29.42,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Hambúrguer bovino 800g,Carne,Chef Express,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Mar & Serra Foods,Portugal,Ativo,Marketplace,Plataformas externas de venda
IT000000026,PED2021011100012,2021-01-11,P020,C0136,F002,CH04,4,18.44,18.44,9.36,9.080000000000002,1,29.42,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Batata smile 1500g,Batatas,FrioMax,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Atlântico Frio Distribuição,Portugal,Ativo,Marketplace,Plataformas externas de venda
IT000000028,PED2021011100013,2021-01-11,P004,C0136,F001,CH02,2,17.44,17.44,9.8,7.640000000000001,1,53.79,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Bacalhau desfiado 800g,Peixe,PT Frozen Foods,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,PT Frozen Foods,Portugal,Ativo,Vendas Externas,Equipa comercial visitando clientes
IT000000030,PED2021011100013,2021-01-11,P023,C0136,F005,CH02,4,11.88,11.28,5.44,5.839999999999999,1,53.79,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Pão de forma 800g,Padaria,Doce Norte,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Doces do Norte,Portugal,Ativo,Vendas Externas,Equipa comercial visitando clientes
IT000000029,PED2021011100013,2021-01-11,P020,C0136,F002,CH02,3,14.79,14.79,6.720000000000001,8.069999999999999,1,53.79,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Batata smile 1500g,Batatas,FrioMax,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Atlântico Frio Distribuição,Portugal,Ativo,Vendas Externas,Equipa comercial visitando clientes
IT000000031,PED2021011100013,2021-01-11,P005,C0136,F003,CH02,1,10.38,10.28,6.99,3.289999999999999,1,53.79,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Miolo de camarão 60/80 800g,Marisco,Norte Mar,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,PortoMar Congelados,Portugal,Ativo,Vendas Externas,Equipa comercial visitando clientes
IT000000036,PED2021011100016,2021-01-11,P012,C0136,F006,CH03,3,17.31,17.31,9.870000000000001,7.439999999999998,1,91.41000000000001,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Hambúrguer bovino 800g,Carne,Chef Express,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Mar & Serra Foods,Portugal,Ativo,Telefone,Pedidos feitos por telefone
IT000000035,PED2021011100016,2021-01-11,P014,C0136,F006,CH03,13,74.10000000000001,74.10000000000001,44.85,29.250000000000007,1,91.41000000000001,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Almôndegas de novilho 1kg,Carne,Mesa Pronta,Inativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Mar & Serra Foods,Portugal,Ativo,Telefone,Pedidos feitos por telefone
IT000000039,PED2021011100018,2021-01-11,P002,C0136,F003,CH01,2,21.72,19.98,13.68,6.300000000000001,1,64.83,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Lombos de salmão 600g,Peixe,Mar Azul,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,PortoMar Congelados,Portugal,Ativo,E-commerce,Pedidos realizados pelo site B2B/B2C
IT000000040,PED2021011100018,2021-01-11,P007,C0136,F002,CH01,5,47.199999999999996,44.85,32.45,12.399999999999999,1,64.83,2021,1,1,11,Janeiro,1,Segunda-feira,0,0,0,Cocktail de marisco 1kg,Marisco,FrioMax,Ativo,Restaurante,Pequeno,Espinho,Aveiro,Alto,Cluster C,Ativo,Atlântico Frio Distribuição,Portugal,Ativo,E-commerce,Pedidos realizados pelo site B2B

In [0]:
# ========================================
# 10. CANDIDATE MART GRAIN TEST
# ========================================

print("[INFO] Testing candidate grain: one record per data_pedido, produto_id, canal_id")

candidate_grouped = (
    df_mart_sales_performance_candidate
    .groupBy("data_pedido", "produto_id", "canal_id")
    .agg(
        F.round(F.sum("quantity_sold"), 2).alias("quantity_sold"),
        F.round(F.sum("gross_sales_amount"), 2).alias("gross_sales_amount"),
        F.round(F.sum("net_sales_amount"), 2).alias("net_sales_amount"),
        F.round(F.sum("total_cost_amount"), 2).alias("total_cost_amount"),
        F.round(F.sum("gross_margin_amount"), 2).alias("gross_margin_amount"),
        F.countDistinct("pedido_id").alias("order_count"),
        F.sum("line_count").alias("line_count"),
        F.round(F.avg("average_order_value"), 2).alias("average_order_value")
    )
)

print(f"[INFO] Candidate grouped mart row count: {candidate_grouped.count():,}")

display(
    candidate_grouped.orderBy("data_pedido", "produto_id", "canal_id")
)

[INFO] Testing candidate grain: one record per data_pedido, produto_id, canal_id
[INFO] Candidate grouped mart row count: 113,010


data_pedido,produto_id,canal_id,quantity_sold,gross_sales_amount,net_sales_amount,total_cost_amount,gross_margin_amount,order_count,line_count,average_order_value
2021-01-11,P002,CH01,2,21.72,19.98,13.68,6.3,1,1,64.83
2021-01-11,P002,CH02,1,11.24,11.02,6.75,4.27,1,1,32.79
2021-01-11,P003,CH02,3,19.74,18.15,11.94,6.21,1,1,18.15
2021-01-11,P004,CH01,2,16.68,16.18,10.18,6.0,1,1,16.18
2021-01-11,P004,CH02,15,125.93,123.34,75.42,47.92,8,10,54.84
2021-01-11,P004,CH03,17,143.07,140.87,83.28,57.59,4,5,53.96
2021-01-11,P005,CH02,20,215.1,208.89,143.44,65.45,6,7,56.07
2021-01-11,P005,CH03,1,10.64,10.32,7.42,2.9,1,1,53.52
2021-01-11,P006,CH03,10,71.72,71.72,41.08,30.64,2,2,41.02
2021-01-11,P006,CH04,1,7.18,6.96,4.16,2.8,1,1,14.04


In [0]:
# ========================================
# 11. CANDIDATE MART VALIDATION
# ========================================

display(
    candidate_grouped.select(
        F.count("*").alias("total_rows"),
        F.countDistinct(
            F.concat_ws("||", F.col("data_pedido"), F.col("produto_id"), F.col("canal_id"))
        ).alias("distinct_grain_keys"),
        F.sum(F.when(F.col("data_pedido").isNull(), 1).otherwise(0)).alias("null_data_pedido"),
        F.sum(F.when(F.col("produto_id").isNull(), 1).otherwise(0)).alias("null_produto_id"),
        F.sum(F.when(F.col("canal_id").isNull(), 1).otherwise(0)).alias("null_canal_id"),
        F.sum(F.when(F.col("net_sales_amount").isNull(), 1).otherwise(0)).alias("null_net_sales_amount"),
        F.sum(F.when(F.col("gross_margin_amount").isNull(), 1).otherwise(0)).alias("null_gross_margin_amount")
    )
)

print("[INFO] Candidate mart validation completed successfully.")

total_rows,distinct_grain_keys,null_data_pedido,null_produto_id,null_canal_id,null_net_sales_amount,null_gross_margin_amount
113010,113010,0,0,0,0,0


[INFO] Candidate mart validation completed successfully.


In [0]:
# ========================================
# 12. METRIC RECONCILIATION CHECK
# ========================================

fact_totals = df_fact_sales.select(
    F.round(F.sum("quantity_sold"), 2).alias("fact_quantity_sold"),
    F.round(F.sum("gross_sales_amount"), 2).alias("fact_gross_sales_amount"),
    F.round(F.sum("net_sales_amount"), 2).alias("fact_net_sales_amount"),
    F.round(F.sum("total_cost_amount"), 2).alias("fact_total_cost_amount"),
    F.round(F.sum("gross_margin_amount"), 2).alias("fact_gross_margin_amount"),
    F.countDistinct("pedido_id").alias("fact_order_count"),
    F.sum("line_count").alias("fact_line_count")
)

candidate_totals = candidate_grouped.select(
    F.round(F.sum("quantity_sold"), 2).alias("mart_quantity_sold"),
    F.round(F.sum("gross_sales_amount"), 2).alias("mart_gross_sales_amount"),
    F.round(F.sum("net_sales_amount"), 2).alias("mart_net_sales_amount"),
    F.round(F.sum("total_cost_amount"), 2).alias("mart_total_cost_amount"),
    F.round(F.sum("gross_margin_amount"), 2).alias("mart_gross_margin_amount"),
    F.sum("order_count").alias("mart_order_count"),
    F.sum("line_count").alias("mart_line_count")
)

display(fact_totals.crossJoin(candidate_totals))

fact_quantity_sold,fact_gross_sales_amount,fact_net_sales_amount,fact_total_cost_amount,fact_gross_margin_amount,fact_order_count,fact_line_count,mart_quantity_sold,mart_gross_sales_amount,mart_net_sales_amount,mart_total_cost_amount,mart_gross_margin_amount,mart_order_count,mart_line_count
914793,5780627.7,5487339.26,3328096.49,2159242.77,89526,244357,914793,5780627.7,5487339.26,3328096.49,2159242.77,216492,244357


In [0]:
# ========================================
# 13. MISSING VALUES ANALYSIS — MART SALES PERFORMANCE
# ========================================

def analyze_missing_values(df, dataset_name):
    print("=" * 80)
    print(f"MISSING VALUES ANALYSIS — {dataset_name.upper()}")
    print("=" * 80)

    total_rows = df.count()

    missing_values_df = df.select([
        F.sum(F.when(F.col(c).isNull(), 1).otherwise(0)).alias(c)
        for c in df.columns
    ])

    missing_values_transposed = (
        missing_values_df
        .select(
            F.array([
                F.struct(
                    F.lit(col_name).alias("column_name"),
                    F.col(col_name).alias("null_count")
                ) for col_name in missing_values_df.columns
            ]).alias("missing_data")
        )
        .select(F.explode("missing_data").alias("data"))
        .select(
            F.col("data.column_name"),
            F.col("data.null_count")
        )
        .withColumn(
            "null_percentage",
            F.round((F.col("null_count") / F.lit(total_rows)) * 100, 4)
        )
        .orderBy(F.col("null_percentage").desc())
    )

    display(missing_values_transposed)

    print(f"[INFO] Total rows analyzed: {total_rows:,}")
    print("[INFO] Missing values analysis completed successfully.")
    print("=" * 80)

analyze_missing_values(candidate_grouped, "mart_sales_performance")

MISSING VALUES ANALYSIS — MART_SALES_PERFORMANCE


column_name,null_count,null_percentage
gross_margin_amount,0,0.0
gross_sales_amount,0,0.0
data_pedido,0,0.0
canal_id,0,0.0
net_sales_amount,0,0.0
produto_id,0,0.0
line_count,0,0.0
order_count,0,0.0
total_cost_amount,0,0.0
quantity_sold,0,0.0


[INFO] Total rows analyzed: 113,010
[INFO] Missing values analysis completed successfully.


%md
### Gold Layer Exploratory Conclusion — mart_sales_performance

This exploratory notebook validated the dataset required to build the `mart_sales_performance` analytical mart in the Gold layer of the PT Frozen Foods data platform. The objective was to confirm data quality, validate aggregation logic, and ensure readiness for analytical consumption.

---

### Grain Validation

The dimensional grain for the `mart_sales_performance` table was confirmed as:

- **One record per `order_date`, `product_id`, and `channel_id`.**

Validation results:

- Total rows: 113,010
- Distinct grain combinations: 113,010
- Null values: 0
- Duplicate values: 0

This confirms that the dataset is unique, consistent, and suitable for analytical modeling.

---

### Business Value Assessment

The `mart_sales_performance` mart enables comprehensive analysis of commercial performance, supporting data-driven decision-making across sales, finance, and operations.

#### High-Value Analytical Metrics

The following measures provide significant business insights:

- **Quantity Sold (`quantity_sold`)**
  - Enables demand analysis and product performance evaluation.

- **Gross Sales Amount (`gross_sales_amount`)**
  - Supports revenue monitoring and trend analysis.

- **Net Sales Amount (`net_sales_amount`)**
  - Provides accurate financial insights after adjustments and discounts.

- **Total Cost Amount (`total_cost_amount`)**
  - Enables cost control and operational efficiency assessments.

- **Gross Margin Amount (`gross_margin_amount`)**
  - Supports profitability and strategic pricing decisions.

- **Line Count (`line_count`)**
  - Measures transactional volume and sales activity.

- **Order Count (`order_count`)**
  - Represents the number of distinct orders within the defined grain.

#### Technical Note on Order Count

The `order_count` metric is classified as **non-additive**. Due to the aggregation level of the mart, a single order may appear across multiple products or channels.

| Source | Value |
|--------|-------|
| Distinct Orders in Fact Table | 89,526 |
| Aggregated Order Count in Mart | 216,492 |

For enterprise-level reporting, the correct calculation of total orders should be performed using:

- `COUNT(DISTINCT order_id)` directly from the `fact_sales` table.

---

### Data Quality Assessment

The dataset demonstrated a high level of quality and consistency.

| Metric | Result |
|--------|--------|
| Total Records | 113,010 |
| Duplicate Records | 0 |
| Null Values | 0 |
| Schema Consistency | Validated |
| Referential Integrity | Preserved |
| Grain Consistency | Confirmed |
| Metric Reconciliation | Validated |

All validations confirm the dataset's reliability for analytical and predictive workloads.

---

### Dimensional Model Alignment

The exploratory analysis confirmed alignment with the dimensional model defined in the project documentation.

Validated entities:

- `mart_sales_performance`
- `fact_sales`
- `dim_product`
- `dim_channel`
- `dim_customer`
- `dim_calendar`

This mart enhances the analytical ecosystem by consolidating key sales metrics for executive reporting and business intelligence initiatives.

Detailed specifications are documented in **12_data_model.md**.

---

### Medallion Architecture Alignment

The dataset is consistent with the Medallion Architecture:

| Layer | Status |
|-------|--------|
| Bronze | Completed |
| Silver | Completed |
| Gold | Validated |

---

### Governance and Technology Alignment

| Component | Technology |
|-----------|------------|
| Storage | ADLS Gen2 |
| Format | Delta Lake |
| Processing | Azure Databricks |
| Governance | Unity Catalog |
| Orchestration | Azure Data Factory |

These standards are defined in the project architecture documentation.

---

### Key Decisions

| Decision | Outcome |
|----------|---------|
| Mart grain | Defined as `order_date`, `product_id`, and `channel_id` |
| Data quality | Fully validated |
| Metric reconciliation | Successfully confirmed |
| Order count behavior | Classified as a non-additive metric |
| Dimensional modeling | Star Schema adopted |
| Referential integrity | Validated and enforced |
| Gold layer readiness | Confirmed |
| Analytical readiness | Approved for BI and reporting |
| Scalability | Designed for future expansion |

---

### Conclusion

The dataset has been validated and approved for Gold layer processing. The analysis ensures:

- data quality and integrity;
- adherence to dimensional modeling best practices;
- reliability of aggregated sales metrics;
- scalability for business intelligence and machine learning applications;
- alignment with enterprise Lakehouse standards.

The platform is now ready for the implementation of the `mart_sales_performance` Gold processing notebook.